### DISCUSSION TITLE
Can Autogen work like OpenAI Function calling?

### DISCUSSION THREAD

**Discussion by WebsheetPlugin** [microsoft/autogen#2208] [Q&A]
I like the concept of Autogen, I like to use a couple of features of it. But I currently just need a simple tool executor.

I notice that Autogen heavily works on the concept of generating code and executing it. Does it not has an simple way of executing code directly after it has been "selected"? - So basically same like OpenAI Function calling? Does it all use function calling?

I ask that, because for my use case it's sometimes enough to just select the correct tool. No need of extra checks etc.

**Comment by jackgerrits** [comment]
Yes it supports OpenAI style tool calls - see [here](https://microsoft.github.io/autogen/docs/tutorial/tool-use)

**Reply by WebsheetPlugin** [reply]
Yes, I noticed that. but this not same as function calling as all their examples still post processes the tool usage with AI again.

Also it seems that in your example, there is one agent which is choosing the tools and the other agent which is using them. So it looks like 3 executions, while openAI function calling would do it with one.

--------------

So if I am not wrong, to execute the tool, it uses:

1. UserProxy (Asks the the Assistant)
2. Assistant (Finds and fills the tool)
(Tool Execution)
3. UserProxy(Writes the response)

--------------

I just want:

1. Assistant (Chooses and fills the tool)
(Tool Execution)

**Reply by jackgerrits** [reply]
In AutoGen, agents generate a response based on their input messages. Therefore, the general pattern is as you described where one agent writes code, or generates a tool call and then another agent responds to that message by processing the message.

It's definitely possible to do as you describe but you'd need to add a [custom reply function](https://microsoft.github.io/autogen/docs/FAQ#how-to-register-a-reply-function) which both generates a tool call, executes it and then emits its response as the executed tool call.

**Reply by x-technik** [reply]
Hello- Which local models do you recommend to use for a simple function call like a schedule check, I do not have any luck with any open source models yet open AI models responds like a butter,

Open AI  gpt 3.5 response-

***** Response from calling tool (call_Ks1SvQnMayTixgPE54lfqOAG) *****
[["10:00", "Client Call"], ["15:00", "Code Review"]]
**********************************************************************

--------------------------------------------------------------------------------
No default IOStream has been set, defaulting to IOConsole.
No default IOStream has been set, defaulting to IOConsole.
No default IOStream has been set, defaulting to IOConsole.
joe (to cathy):

Your schedule for April 2, 2024 is as follows:
- 10:00 AM: Client Call
- 3:00 PM: Code Review

**Reply by sonichi** [reply]
One can create an agent that's able to both propose and execute tools, using a nested chat. cc @qingyun-wu 
I've suggested this a few times. Let's create an issue.

**Reply by ekzhu** [reply]
Check out the chess example for an agent that can execute tool internally. https://microsoft.github.io/autogen/docs/notebooks/agentchat_nested_chats_chess

**Reply by WebsheetPlugin** [reply]
Yes, I also noticed this now. Thanks for your help although. I might just add that, inside the notebooks I missing the most simple way to redo the functionality of openai function calling, and for me, it would be like you suggested to use the same agent for register_for_llm and register_for_execution.

As this proxy logic is a little complex at the start and hard to understand at the beginning.

**Reply by ekzhu** [reply]
Agree. A simpler higher-level agent like GPTAssistantAgent could be helpful here. 

GPTAssistantAgent: https://microsoft.github.io/autogen/docs/reference/agentchat/contrib/gpt_assistant_agent
